### PR TITLE
Add Serverless Run Plugin

### PR DESCRIPTION
**Pull request by eahefnawy**
## What did you implement:

`serverless run` command that installs and/or downloads the event gateway and the local emulator, spins them up and deploys & registers functions to the local emulator and/or event gateway.

## How did you implement it:

We first check whether the local emulator is running, if it is, we deploy functions to the local emulator. We then check whether the event gateway is running, if it is, we register functions/events to the event gateway. If they're not running, we check whether they're installed and install them if not. We then spin them up and deploy and register functions.

#### Basic Example

- Clone the basic example from the [examples repo](https://github.com/serverless/event-gateway-examples)
- from inside the service directory, run `sls run`
- After functions are deployed/registered, visit `localhost:4000/e1`. You should see the response from the function.

## Todos

- [x] Publish Local Emulator
- [x] Fix FDK.configure bug
- [x] Subscribe functions to events
- [ ] Support Google Cloud Functions
- [x] Fix linting errors
- [ ] cors: true makes the gateway error. we should not pass it to the gateway and use a property whitelisting approach
- [x] make sure to uppercase method values (will be solved in the event gateway)
- [ ] make sure showing the data in serverless run output works with other dataTypes than JSON
- [x] improve emit terminal output
- [ ] we split the lines by \n. makes sure this doesn't affect escaped newlines inside event content

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

### CODE REVIEW

**Comment by rupakg** [comment]
@eahefnawy I gave it a spin and I got errors in installing the `serverless-local-emulator`. Looks like `serverless-local-emulator` is not yet published to npm.

`npm http 404 https://registry.npmjs.org/serverless-local-emulator`

Full trace:

```
root@0555e2cc096a:/home/prjtest/apps/email-service# sls run
 Serverless     |  Installing Local Emulator...
npm info it worked if it ends with ok
npm info using npm@3.10.10
npm info using node@v6.10.3
npm info attempt registry request try #1 at 6:33:19 PM
npm http request GET https://registry.npmjs.org/serverless-local-emulator
npm http 404 https://registry.npmjs.org/serverless-local-emulator
npm ERR! Linux 4.9.31-moby
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "install" "-g" "serverless-local-emulator"
npm ERR! node v6.10.3
npm ERR! npm  v3.10.10
npm ERR! code E404

npm ERR! 404 Registry returned 404 for GET on https://registry.npmjs.org/serverless-local-emulator
npm ERR! 404
npm ERR! 404  'serverless-local-emulator' is not in the npm registry.
npm ERR! 404 You should bug the author to publish it (or use the name yourself!)
npm ERR! 404
npm ERR! 404 Note that you can also install from a
npm ERR! 404 tarball, folder, http url, or git url.

npm ERR! Please include the following file with any support request:
npm ERR!     /home/prjtest/apps/email-service/npm-debug.log

  Error --------------------------------------------------

  Command failed: npm install -g serverless-local-emulator
npm info it worked if it ends with ok
npm info using npm@3.10.10
npm info using node@v6.10.3
npm info attempt registry request try #1 at 6:33:19 PM
npm http request GET https://registry.npmjs.org/serverless-local-emulator
npm http 404 https://registry.npmjs.org/serverless-local-emulator
npm ERR! Linux 4.9.31-moby
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "install" "-g" "serverless-local-emulator"
npm ERR! node v6.10.3
npm ERR! npm  v3.10.10
npm ERR! code E404

npm ERR! 404 Registry returned 404 for GET on https://registry.npmjs.org/serverless-local-emulator
npm ERR! 404
npm ERR! 404  'serverless-local-emulator' is not in the npm registry.
npm ERR! 404 You should bug the author to publish it (or use the name yourself!)
npm ERR! 404
npm ERR! 404 Note that you can also install from a
npm ERR! 404 tarball, folder, http url, or git url.

npm ERR! Please include the following file with any support request:
npm ERR!     /home/prjtest/apps/email-service/npm-debug.log


     For debugging logs, run again after setting the "SLS_DEBUG=*" environment variable.

  Get Support --------------------------------------------
     Docs:          docs.serverless.com
     Bugs:          github.com/serverless/serverless/issues
     Forums:        forum.serverless.com
     Chat:          gitter.im/serverless/serverless

  Your Environment Information -----------------------------
     OS:                     linux
     Node Version:           6.10.3
     Serverless Version:     1.19.0
```

**Comment by pmuens** [comment]
@rupakg thanks for testing! 👍

AFAIK it's now called @serverless/emulator and is already published to `npm`. @eahefnawy just needs to update the code here (which he already has on his Todo list for this PR AFAIK).

**Comment by rupakg** [comment]
@pmuens Maybe I can do a PR.

**Comment by pmuens** [comment]
> @pmuens Maybe I can do a PR. But, I could not find it on npm.

Thanks for that @rupakg 👍

🤔 maybe @eahefnawy knows more...

**Comment by eahefnawy** [comment]
local emulator not published yet, gonna publish it today

**Comment by rupakg** [comment]
@eahefnawy @pmuens Any word on resolving this issue? 

`serverless run` makes it easy to use `local-emulator` and `event-gateway` altogether with the framework. Thanks for working on it. I am kinda stuck with this issue...

**Comment by pmuens** [comment]
> @eahefnawy @pmuens Any word on resolving this issue?

> serverless run makes it easy to use local-emulator and event-gateway altogether with the framework. Thanks for working on it. I am kinda stuck with this issue...

The setup to auto-deploy the local-emulator package to `npm` should be in place so I think it should be published in the beginning of the next week.

**Comment by rupakg** [comment]
@eahefnawy I got latest of the branch and tried it on a service I have. I get this error:

```
npm info ok
 Serverless     |  Installing Event Gateway...
 Serverless     |  Spinning Up the Local Emulator...
 Serverless     |  Spinning Up the Event Gateway...
 Local Emulator |  /usr/local/lib/node_modules/@serverless/emulator/node_modules/koa-bodyparser/index.js:72
  return async function bodyParser(ctx, next) {
               ^^^^^^^^
SyntaxError: Unexpected token function
    at createScript (vm.js:56:10)
    at Object.runInThisContext (vm.js:97:10)
    at Module._compile (module.js:542:28)
    at Object.Module._extensions..js (module.js:579:10)
    at Module.load (module.js:487:32)
    at tryModuleLoad (module.js:446:12)
    at Function.Module._load (module.js:438:3)
    at Module.require (module.js:497:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/usr/local/lib/node_modules/@serverless/emulator/dist/main.js:211:22)
 Event Gateway  |  INFO	Running in development mode with embedded etcd.
 Serverless     |  Registering Functions to the Event Gateway...
 Event Gateway  |  INFO	Function registered.	{"functionId": "email-service-send", "type": "http"}

  Error --------------------------------------------------

  Failed to register at least one function.


     For debugging logs, run again after setting the "SLS_DEBUG=*" environment variable.

  Get Support --------------------------------------------
     Docs:          docs.serverless.com
     Bugs:          github.com/serverless/serverless/issues
     Forums:        forum.serverless.com
     Chat:          gitter.im/serverless/serverless

  Your Environment Information -----------------------------
     OS:                     linux
     Node Version:           6.10.3
     Serverless Version:     1.19.0
```

Any ideas?

**Comment by pmuens** [comment]
It looks like the `node_modules` in the `local-emulator` (`koa` in that case) are not transpiled correctly so they won't work with early node versions.

/cc @eahefnawy

File: lib/plugins/run/utils/installEventGateway.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
+'use strict';
+
+const download = require('download');
+const os = require('os');
+const path = require('path');
+
+function installEventGateway(eventGatewayVersion) {
+  let eventGatewayDownloadUrl = `https://github.com/serverless/event-gateway/releases/download/${eventGatewayVersion}/event-gateway_${eventGatewayVersion}_darwin_386.tar.gz`;
```
Comment (mthenw) [inline lib/plugins/run/utils/installEventGateway.js]: can we download `amd64` version for 64bit arch?

Comment (nikgraf) [inline lib/plugins/run/utils/installEventGateway.js]: I think this could do it https://stackoverflow.com/a/37465917/837709

**Comment by rupakg** [comment]
@eahefnawy @pmuens I tried the `run` command with the latest update, and I am running into this issue:

```
npm info ok
 Serverless     |  Installing Event Gateway...
 Serverless     |  Spinning Up the Local Emulator...
 Serverless     |  Spinning Up the Event Gateway...
 Event Gateway  |  INFO	Running in development mode with embedded etcd.
 Local Emulator |  Tue, 08 Aug 2017 20:13:55 GMT koa deprecated Support for generators will be removed in v3. See the documentation for examples of how to convert old middleware https://github.com/koajs/koa/blob/master/docs/migration.md at ../../../../../usr/local/lib/node_modules/@serverless/emulator/dist/main.js:14:17
 Local Emulator |  Serverless Local Emulator Daemon listening at localhost:4002...
 Serverless     |  Deploying Functions to Local Emulator...
 Serverless     |  Functions Deployed to the Local Emulator!
 Serverless     |  Registering Functions to the Event Gateway...
 Event Gateway  |  DEBUG	Function registered.	{"functionId": "email-service-send", "type": "http"}

  Error --------------------------------------------------

  Failed to register at least one function.


     For debugging logs, run again after setting the "SLS_DEBUG=*" environment variable.

  Get Support --------------------------------------------
     Docs:          docs.serverless.com
     Bugs:          github.com/serverless/serverless/issues
     Forums:        forum.serverless.com
     Chat:          gitter.im/serverless/serverless

  Your Environment Information -----------------------------
     OS:                     linux
     Node Version:           6.10.3
     Serverless Version:     1.19.0
```

**Highlighting the error:**

```
Event Gateway  |  DEBUG	Function registered.	{"functionId": "email-service-send", "type": "http"}

  Error --------------------------------------------------

  Failed to register at least one function.
```

Another couple of things from the output:

**Can we get rid of this depreciation message?**

```
 Local Emulator |  Tue, 08 Aug 2017 20:13:55 GMT koa deprecated Support for generators will be removed in v3. See the documentation for examples of how to convert old middleware https://github.com/koajs/koa/blob/master/docs/migration.md at ../../../../../usr/local/lib/node_modules/@serverless/emulator/dist/main.js:14:17
```

**I really like how the Local Emulator advertises the port it listen's on.** I would like the Event Gateway to do the same. ([Issue 246](https://github.com/serverless/event-gateway/issues/246))

```
 Local Emulator |  Serverless Local Emulator Daemon listening at localhost:4002...
```

**Maybe at the bottom, we can also show version info. for Local Emulator and Event Gateway as well:**

```
  Your Environment Information -----------------------------
     OS:                     linux
     Node Version:           6.10.3
     Serverless Version:     1.19.0
```

**PS**: I was not sure if opening issues for these would confuse existing users as the feature is not merged in yet.

**Comment by nikgraf** [comment]
Shout out to @eahefnawy 🙌 The code is clean and easy to work with!

**Comment by rupakg** [comment]
@eslam This is the latest error that I see:

```
npm info ok
 Serverless     |  Installing Event Gateway
 Serverless     |  Spinning Up the Local Emulator
 Serverless     |  Spinning Up the Event Gateway
undefined:2
{"level":"debug","ts":1502379711.8051083,"msg":"Function local cache received value update.","key":"users-service-registerUser","value":"{\"functionId\":\"users-service-registerUser\",\"provider\":{\"type\":\"http\",\"url\":\"http://localhost:4002/v0/emulator/api/invoke/users-service/registerUser\"}}"}
^

SyntaxError: Unexpected token { in JSON at position 122
    at Object.parse (native)
    at module.exports.msg (/usr/local/lib/node_modules/serverless/lib/plugins/run/utils/logEventGateway.js:40:26)
    at Socket.eventGatewayChildProcess.stderr.on.chunk (/usr/local/lib/node_modules/serverless/lib/plugins/run/index.js:155:15)
    at emitOne (events.js:96:13)
    at Socket.emit (events.js:188:7)
    at readableAddChunk (_stream_readable.js:176:18)
    at Socket.Readable.push (_stream_readable.js:134:10)
    at Pipe.onread (net.js:547:20)
```

**Comment by rupakg** [comment]
@eahefnawy The above issue is resolved thanks to @nikgraf

**Comment by rupakg** [comment]
@eahefnawy @pmuens @nikgraf 

Here is the new issue:

First service:

```
$ cd ~/projects/svrless/apps/event-gateway-example/services/users-service
$ sls run
 Serverless     |  Installing Local Emulator
 Serverless     |  Installing Event Gateway
 Serverless     |  Spinning Up the Local Emulator
 Serverless     |  Spinning Up the Event Gateway
 Event Gateway  |  Running in development mode with embedded etcd. Event API listening on http://localhost:4000. Config API listening on http://localhost:4001.
 Event Gateway  |  {"level":"debug","ts":1502391041.2814264,"msg":"Subscription created.","event":"http","functionId":"users-service-registerUser"}
{"level":"debug","ts":1502391041.2814353,"msg":"Subscription created.","event":"http","functionId":"users-service-getUser"}

 Local Emulator |  Serverless Local Emulator Daemon listening at localhost:4002...
```
Here in the logs, shouldn't it log that the two functions `registerUser` and `getUser` are registered with EG? Like this:

```
DEBUG    Function registered.    {"functionId": "registerUser", "type": "awslambda"}
DEBUG    Function registered.    {"functionId": "getUser", "type": "awslambda"}
DEBUG    Function registered.    {"functionId": "sendWelcomeEmail", "type": "awslambda"}
```

Second service:

```
$ cd ~/projects/svrless/apps/event-gateway-example/services/users-service
$ sls run
 Serverless     |  Local Emulator Already Running
 Serverless     |  Event Gateway Already Running
 Serverless     |  Installing Event Gateway
```
It just exits out.

Any ideas?

**Comment by aheissenberger** [comment]
I am using "serverless" **installed local with the project** and would like to do this with the emulator too.
But whenever I call `../../node_modules/.bin/sls run` it tries to do a global install and fails on my system as I do not have admin rights:
```
$ ../../node_modules/.bin/sls run
 Serverless     Installing Emulator
npm WARN checkPermissions Missing write access to /usr/local/lib/node_modules
npm ERR! path /usr/local/lib/node_modules
npm ERR! code EACCES
npm ERR! errno -13
npm ERR! syscall access
npm ERR! Error: EACCES: permission denied, access '/usr/local/lib/node_modules'
npm ERR!  { Error: EACCES: permission denied, access '/usr/local/lib/node_modules'
npm ERR!   stack: 'Error: EACCES: permission denied, access \'/usr/local/lib/node_modules\'',
npm ERR!   errno: -13,
npm ERR!   code: 'EACCES',
npm ERR!   syscall: 'access',
npm ERR!   path: '/usr/local/lib/node_modules' }
npm ERR! 
npm ERR! Please try running this command again as root/Administrator.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/ah/.npm/_logs/2017-09-20T19_38_14_739Z-debug.log
 
  Error --------------------------------------------------
 
  Command failed: npm install -g @serverless/emulator@0.1.18
npm WARN checkPermissions Missing write access to /usr/local/lib/node_modules
npm ERR! path /usr/local/lib/node_modules
npm ERR! code EACCES
npm ERR! errno -13
npm ERR! syscall access
npm ERR! Error: EACCES: permission denied, access '/usr/local/lib/node_modules'
npm ERR!  { Error: EACCES: permission denied, access '/usr/local/lib/node_modules'
npm ERR!   stack: 'Error: EACCES: permission denied, access \'/usr/local/lib/node_modules\'',
npm ERR!   errno: -13,
npm ERR!   code: 'EACCES',
npm ERR!   syscall: 'access',
npm ERR!   path: '/usr/local/lib/node_modules' }
npm ERR! 
npm ERR! Please try running this command again as root/Administrator.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/ah/.npm/_logs/2017-09-20T19_38_14_739Z-debug.log

 
     For debugging logs, run again after setting the "SLS_DEBUG=*" environment variable.
 
  Get Support --------------------------------------------
     Docs:          docs.serverless.com
     Bugs:          github.com/serverless/serverless/issues
     Forums:        forum.serverless.com
     Chat:          gitter.im/serverless/serverless
 
  Your Environment Information -----------------------------
     OS:                     darwin
     Node Version:           8.5.0
     Serverless Version:     1.22.0
```

packages.json
```
"devDependencies": {
   "serverless": "^1.22.0",
    "@serverless/emulator": "0.1.18"
  }
```

**Comment by pmuens** [comment]
@aheissenberger thanks for commenting 👍

🤔 nice catch! It looks like the Emulator and Event Gateway support is only available in global installs.

Could you please open up a separate issue so that we can track and discuss this in isolation? Thanks in advance!

**Pull request merged**